File: public/packages/typescript/resilience/src/pipeline/i-resilient-layer.ts
// -----------------------------------------------------------------------
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) DCSV
// -----------------------------------------------------------------------

/**
 * One layer in a {@link ResilientPipeline}. Wraps the inner async op with its
 * own resilience concern (singleflight / breaker / retry / timeout / rate-limit)
 * and delegates downward.
 *
 * Mirrors .NET `IResilientLayer<TKey, TValue>.WrapAsync(key, next, ct)`: the
 * caller threads an optional {@link AbortSignal} (≈ `CancellationToken`) and the
 * op receives the signal it should observe for cooperative cancellation. A layer
 * may replace the signal it passes inward (e.g. {@link TimeoutLayer} hands the op
 * a signal linked to both the caller signal and its own timeout) — exactly as the
 * .NET `TimeoutLayer` passes a linked `CancellationToken` down.
 */
export interface IResilientLayer {
  /**
   * @param key Per-call key (used by Singleflight; ignored by other layers).
   * @param op  The inner op. Receives the {@link AbortSignal} it should observe
   *            (may be `undefined` when no cancellation is in play).
   * @param signal Optional caller cancellation signal threaded down the stack.
   */
  execute<T>(
    key: string,
    op: (signal?: AbortSignal) => Promise<T>,
    signal?: AbortSignal,
  ): Promise<T>;
}

File: public/packages/dotnet/resilience/Pipeline/IResilientLayer.cs
// -----------------------------------------------------------------------
// <copyright file="IResilientLayer.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Resilience.Pipeline;

/// <summary>
/// A single decoration step in a <see cref="ResilientPipeline{TKey, TValue}"/>.
/// Each layer wraps the next inner callable, optionally consulting the
/// per-call key (used by Singleflight; ignored by other layers).
/// </summary>
/// <typeparam name="TKey">
/// Per-call key type. Used by layers that dedupe / route by key
/// (Singleflight); other layers receive but ignore it.
/// </typeparam>
/// <typeparam name="TValue">The value type produced by the operation.</typeparam>
public interface IResilientLayer<TKey, TValue>
    where TKey : notnull
{
    /// <summary>
    /// Wraps <paramref name="next"/> with this layer's behavior and invokes it.
    /// </summary>
    /// <param name="key">The per-call key (used by some layers, ignored by others).</param>
    /// <param name="next">The inner callable to wrap.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The (possibly transformed) operation result.</returns>
    ValueTask<TValue> WrapAsync(
        TKey key,
        Func<CancellationToken, ValueTask<TValue>> next,
        CancellationToken ct);
}
